src/main/java/org/apache/sysml/hops/BinaryOp.java

```diff
@@ -920,6 +920,8 @@ else if( op == OpOp2.RBIND ) {
 				return new long[]{ldim1, ldim2, lnnz};
 		}
 		else if ( op == OpOp2.SOLVE ) {
+			if (mc[0].getRows() != mc[0].getCols())
+				throw new RuntimeException("The A matrix, in solve(A,b) should have squared dimensions.");
```

Comment:
Could you move this check please into (1) the builtin expression validation, and (2) the runtime operation to ensure the error is always thrown independent of the size propagation being able to infer the size or not. Also not that this method for inferring worst-case output sizes is only executed if we cannot exactly infer the output sizes.